src/main/java/me/ryanhamshire/GPFlags/util/ChatUtil.java

```diff
@@ -0,0 +1,258 @@
+            case '5':
+                colorCode = "<color:dark_purple>";
+            case '6':
+                colorCode = "<color:gold>";
+            case '7':
+                colorCode = "<color:gray>";
+            case '8':
+                colorCode = "<color:dark_gray>";
+            case '9':
+                colorCode = "<color:blue>";
+            case 'l':
+                colorCode = "<bold>";
+            case 'm':
+                colorCode = "<strikethrough>";
+            case 'n':
+                colorCode = "<u>";
+            case 'o':
+                colorCode = "<italic>";
+            case 'k':
+                colorCode = "<obf>";
+        }
+        switch (color) {
+            case 'l':
+            case 'm':
+            case 'n':
+            case 'o':
+            case 'k':
+                break;
+            default:
+                colorCode = reset + colorCode;
+        }
+        return colorCode;
+    }
+
+    private static String validateMinimessage(CommandSender sender, String message) {
+
+        // Check if there is a link in the message without click-component
+        Pattern pattern = Pattern.compile("https?://\\S+");
+        Matcher matcher = pattern.matcher(message);
+        while (matcher.find()) {
+            boolean requiredChanges = false;
+            if(!sender.hasPermission("gpflags.messages.links")) {
+                message = message.replace(matcher.group(), "");
+                requiredChanges = true;
+            } else {
+                String link = message.substring(matcher.start(), matcher.end());
+                message = message.replace(link, "<click:open_url:" + link + ">" + link + "</click>");
+                matcher = pattern.matcher(message);
+            }
+            if (!requiredChanges) {
+                break;
+            }
+        }
+
+        // Check for click-components
+        pattern = Pattern.compile("<click:.*?>");
+        matcher = pattern.matcher(message);
+        while (matcher.find()) {
+            boolean requiredChanges = false;
+            String clickComponent = message.substring(matcher.start(), matcher.end());
+            if (clickComponent.contains("suggest_command") && !sender.hasPermission("gpflags.messages.click.suggest_command")) {
+                message = message.replace(clickComponent, "");
+                requiredChanges = true;
+            }
+            if (clickComponent.contains("run_command") && !sender.hasPermission("gpflags.messages.click.run_command")) {
+                message = message.replace(clickComponent, "");
+                requiredChanges = true;
+            }
+            if (clickComponent.contains("open_url") && !sender.hasPermission("gpflags.messages.click.open_url")) {
+                message = message.replace(clickComponent, "");
+                requiredChanges = true;
+            }
+            matcher = pattern.matcher(message);
+
+            if (!requiredChanges) {
```

Comment:
I can't tell what this variable is trying to do in this one. If I don't have permission for a click action I tried to use, it removes my click component, but leaves all the other click components in the message?

Reply:
The problem was, when you find something in the pattern, **but** have the permission to use it, there is no need for change. But if there is no need for change, the while loop will keep iterating over and over until the server runs out of memory.

Reply:
But we could make the permission check before the actually pattern initialization. However it could become annoying with the click and hover components. What do you think about this?

Reply:
Yeah, I think it makes sense for perm check before pattern initialization. What makes it annoying with click and hover components? That there are multiple types of them? Thats fine, just split it up more
```java
if (!sender.hasPermission("gpflags.messages.hover.show_text")) {
// replace all "<hover:show_text>" with "\<hover:show_text>"
}
if (!sender.hasPermission("gpflags.messages.hover.show_item")) {
// replace all "<hover:show_item>" with "\<hover:show_item>"
}
...
```
As you see in my example above, I also chose to escape out the tag rather than repalcing it with "". 
I think if you choose to replace it with "", it could result in some cases where for example \<re\<blue>d> would turn into \<red> which we want to avoid.

Reply:
Dont get this, Im sorry.

Do you want to put a `\` before the <hover:....> stuff? I think this would still setup the component itself. Also, we are having a while loop that repeats until there is no component left. So i think thats not bad at all atm.

All i could imagine to reduce iteration time:
```java
if(!sender.hasPermission("gpflags.messages.hover.show_text") || !sender.hasPermission("gpflags.messages.hover.show_item") || !sender.hasPermission("gpflags.messages.hover.show_entity")) {
pattern = Pattern.compile("<hover:.*?>");
        matcher = pattern.matcher(message);
        while (matcher.find()) {
            boolean requiredChanges = false;

            String hoverComponent = message.substring(matcher.start(), matcher.end());
            if (hoverComponent.contains("show_text") && !sender.hasPermission("gpflags.messages.hover.show_text")) {
                message = message.replace(hoverComponent, "");
                requiredChanges = true;
            }
            if (hoverComponent.contains("show_item") && !sender.hasPermission("gpflags.messages.hover.show_item")) {
                message = message.replace(hoverComponent, "");
                requiredChanges = true;
            }
            if (hoverComponent.contains("show_entity") && !sender.hasPermission("gpflags.messages.hover.show_entity")) {
                message = message.replace(hoverComponent, "");
                requiredChanges = true;
            }
            matcher = pattern.matcher(message);

            if (!requiredChanges) {
                break;
            }
        }
}
```

Reply:
> Do you want to put a \ before the hover:.... stuff? I think this would still setup the component itself. 

Yes that's exactly what Im saying. You can use the minimessage viewer at https://webui.advntr.dev/ to see that this does make the tag show without it serving as a tag.